### DISCUSSION TITLE
Autonomous Navigation Tutorial Feedback

### DISCUSSION THREAD

**Discussion by schmittlema** [prl-mushr/mushr#90] [Software]
This thread is for feedback and questions on the autonomous navigation tutorial

**Comment by Lilly-yang** [comment]
Hi Matt,
Could you please give me access to clone these two repositories? Thank you!
```
# Install the Global Planner (this is for Jetson Xavier NX or sim use)
$ git clone git@github.com:prl-mushr/mushr_gp.git
# If you're using the jetson nano, use the mushr_gprm planner:
$ git clone git@github.com:prl-mushr/mushr_gprm.git
```

**Reply by schmittlema** [reply]
[mushr_gp](https://github.com/prl-mushr/mushr_gp) should now be available. mushr_gprm is still in progress and not ready for an alpha release yet. It should be available in the next week or 2

**Reply by hardikbvaghela** [reply]
# Install the Global Planner (this is for Jetson Xavier NX or sim use)
$ git clone git@github.com:prl-mushr/mushr_gp.git
# If you're using the jetson nano, use the mushr_gprm planner:
$ git clone git@github.com:prl-mushr/mushr_gprm.git

so i am using jatson nano so mushr_gprm is available?? because i already run this but still need git hub username or public key. so now what should i do?? how can i access?

**Reply by MarkusSchiffer** [reply]
Hello, mushr_gprm is a bit behind schedule, and is not out yet. Please keep using mushr_gp for a bit longer. mushr_gprm is almost ready to go.

**Reply by MarkusSchiffer** [reply]
Hello, mushr_gprm is live! Feel free to give it a try.

**Reply by hardikbvaghela** [reply]
but it need public key so what should i do?

**Reply by Lilly-yang** [reply]
> but it need public key so what should i do?

No need. You can try `git clone https://github.com/prl-mushr/mushr_gprm.git`.

**Comment by Lilly-yang** [comment]
Hi Matt,

I got stuck at this step
`[INFO] [1664812273.961424]: Generating 62 trajectories for control, this will take a minute`

Could you please give me some advice?

The map size is 202 x 76.
![image](https://user-images.githubusercontent.com/39558023/193626246-5d94984f-7ccc-4235-9ab2-dddfce610754.png)

Thanks!

**Reply by sidtalia** [reply]
Hi! Apologies for the late reply. Could you switch the branch from noetic to "noetic_rhc_replace"? The docker installation system installs the wrong branch for the mushr_rhc at the moment. This issue has been noted and will be fixed soon.

**Reply by Lilly-yang** [reply]
Hi naughtyStark, thank you! This problem has been solved.

Could you please tell me where the parameter `invert_servo: true` of `vesc.yaml` is read?

I am trying to launch mushr_sim, mushr_rhc sim, and mushr_gprm sim on the robot, but sending control commands to both the sim_car and the real_car at the same time. Everything goes well in the sim but the real car runs reversed: going forward is going backward, and turning left is turning right. I noticed that mushr_base inverts the servo but I didn't find where(which file) read this parameter.
Thx.

**Reply by sidtalia** [reply]
@Lilly-yang, Hi, at the moment the MuSHR stack does not allow for the simultaneous operation of both sim and real, which appears to be what you are trying to do. This limitation exists because of legacy reasons and may not be removed soon (unless it affects all users).

However, there is a fairly easy workaround for what you are trying to achieve. I would recommend not editing the MuSHR stack itself because that may have unintended consequences for the state estimator and so on. I would also like to clarify that the parameter ``` invert_servo```, while defined in the yaml is not actually used anywhere. The inversion is taken care of by the negative sign on the ```steering_angle_to_servo_gain``` parameter. 

Here is my recommendation:
1) Create a python script that subscribes to the following topics:
```
commands/motor/speed
commands/servo/position
```
Make this python script invert the sign / adjust the scaling factor according to the difference in the vesc.yaml found in the mushr_base/vesc and the yaml found in mushr_sim/vesc. Publish the outputs to new topic names:
```
real_commands/motor/speed
real_commands/servo/position
```

2) Remap the input topics for [vesc_driver_node](https://github.com/prl-mushr/vesc/blob/noetic/vesc_main/launch/vesc.launch#L24) so that it takes the "real_command" topics and maps them to "command" topics. 

The idea here is to isolate the control inversion so that it doesn't break the rest of the system. Let me know if this works for you!
